modules/GraphvizLayout/src/main/java/org/icculus/chunky/gephigraphviz/GraphvizLayout.java

```diff
@@ -0,0 +1,314 @@
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+
+        Gary Briggs <chunky@icculus.org>
+*/
+
+package org.icculus.chunky.gephigraphviz;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.MissingResourceException;
+import java.util.Set;
+
+import javax.swing.JOptionPane;
+
+import org.gephi.graph.api.Edge;
+import org.gephi.graph.api.Graph;
+import org.gephi.graph.api.Node;
+// import org.gephi.graph.api.NodeData;
+import org.gephi.layout.plugin.AbstractLayout;
+import org.gephi.layout.spi.Layout;
+import org.gephi.layout.spi.LayoutBuilder;
+import org.gephi.layout.spi.LayoutProperty;
+import org.openide.util.Exceptions;
+import org.openide.util.NbBundle;
+
+public class GraphvizLayout extends AbstractLayout implements Layout {
+
+    // http://www.graphviz.org/doc/info/attrs.html
+    private String algoName = "dot";
+    private String dotBinary = "dot";
+    private String rankDir = "LR";
+    private String overlap = "false";
+    private Boolean concentrate = false;
+    
+    private Process dotprocess = null;
+
+    private Graph graph;
+
+    public GraphvizLayout(LayoutBuilder layoutBuilder) {
+        super(layoutBuilder);
+    }
+
+    @Override
+    public void initAlgo() {
+        this.graph = graphModel.getGraphVisible();
+        setConverged(false);
+    }
+
+    @Override
+    public void goAlgo() {
+        // Prepare input
+        final StringBuffer dotfile = new StringBuffer();
+
+        dotfile.append("digraph g {\n");
+        dotfile.append("layout = \"").append(this.algoName).append("\";\n");
+        dotfile.append("rankdir = \"").append(this.rankDir).append("\";\n");
+        dotfile.append("overlap = \"").append(this.overlap).append("\";\n");
+        if (this.concentrate) {
+            dotfile.append("concentrate=true;\n");
+        }
+
```

Comment:
In order to guarantee consistency I would recommend to add a `graph.readLock()` here and add a `graph.readUnlock()` at line 107. Indeed, the graph could be modified between your node and edge loop, so you could have edges of new/missing nodes for instance.